### DISCUSSION TITLE
Idea for simpler oauth scopes interface

### DISCUSSION THREAD

**Discussion by MatthewScholefield** [fastapi-users/fastapi-users#589] [Ideas]
Hi, first thanks again for all the work you've done with fastapi-users 😄.

So, I was working with OAuth scopes and noticed they are passed via url parameters from the frontend to the authorize endpoint. For example:

```
curl -X http://localhost:5000/auth/google/authorize?authentication_backend=cookie&scopes=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+
```

To me, this seems a little clunky because oauth scopes seem like they are very specific to the backend app and need to be set to a specific default for each backend. For this reason, I was wondering about the reasoning behind the current design and if it could be made simpler. 

Specifically, I was thinking that instead of exposing scopes as url parameters, default scopes from `oauth_client.base_scopes` would be applied automatically and any extra scopes could be passed as arguments within`get_oauth_router`. So something like this:

```python
...
fastapi_users.get_oauth_router(github_oauth_client, secret, after_register=..., scopes=['user:extra-google-scope'])
...
```
```bash
curl http://localhost:5000/auth/github/authorize
```

Let me know your thoughts on this.

**Comment by frankie567** [comment]
Thank you for your feedback!

The reason behind this choice is that it allows you to make incremental scope requests, which is [recommended by Google](https://developers.google.com/identity/protocols/oauth2/web-server#incrementalAuth). It means that you can gradually ask for more scopes when your user needs to use additional features. So, setting scopes dynamically in the query parameters allow this.

That said, there is probably a default behiavor to implement indeed. I probably had something in mind when I defined the `base_scopes` property on the OAuth client, but I think I lost myself along the way.

Here is what we could do:

* I want to keep the query parameters behavior for incremental authorization
* In `httpx-oauth`, rename on the OAuth2 client the property `base_scopes` (which is unused anyway 🤷‍♂️) to simply `scope` and use it as a default in `get_authorization_url` if the `scope` argument is not set
* This way, the `/authorize` route could be called without the `scope` query parameter. And the developers could still set their own set of default scopes by overloading it on the OAuth2 client

What do you think?

**Reply by MatthewScholefield** [reply]
I don't fully follow, but as long as empty query parameters on `/authorize` apply either default or a customizable set of scopes, I think that's reasonable.

And that makes a lot of sense why accepting them as query parameters is useful.

As a side note, I noticed the default scopes for github were `user` and `user:email` but in reality I think it probably should only be `user:email`; when I specified both it asked for entire access to all private info which wouldn't be a good default.

**Reply by frankie567** [reply]
I'll look into it!

As for the GitHub scope, it's actually to fix issue #511 (that you reported by the way 😁): without it, you won't have access to the e-mail address if the user does not display it publicly.

**Reply by MatthewScholefield** [reply]
Oh, I'm actually just referring to the `user` scope (the first of [the two items listed here](https://github.com/frankie567/httpx-oauth/blob/master/httpx_oauth/clients/github.py#L11)). I tested and it can still read my private email even if I don't include the `user` scope (only `user:email`).
